File: src/mcts/mcts.py
from src.mcts.node import Node
import random
import time


class MonteCarloTreeSearch:
    def __init__(self) -> None:
        pass

    def best_move(self, root: Node, simulations_number: int) -> int:
        start = time.time()

        for _ in range(simulations_number):
            node = self._selection(root)
            expanded_node = self._expansion(node)
            result = self._simulation(expanded_node)
            self._backpropagation(expanded_node, result)
        best_child = max(root.get_children(), key=lambda child: (child.simulations, child.wins))

        for child in root.get_children():
            print(child.move, child.wins, child.simulations)
        print()
        print("Time:", time.time() - start)

        return best_child.move

    def _selection(self, root: Node) -> Node:
        node = root
        while node.children:
            node = node.best_child()
        return node

    def _expansion(self, selected_node: Node) -> Node:
        if selected_node.is_terminal():
            return selected_node

        children = selected_node.get_children()
        unexplored_children = [child for child in children if child.simulations == 0]
        return random.choice(unexplored_children)

    def _simulation(self, expanded_node: Node) -> int:
        simulated_game = expanded_node.game.copy()
        while not simulated_game.is_game_over():
            simulated_game.make_random_move()
        return simulated_game.get_winner()

    def _backpropagation(self, terminal_node: Node, result: int) -> None:
        node = terminal_node
        while node is not None:
            node.wins += abs(result) if node.game.current_player != result else -abs(result)
            node.simulations += 1
            node = node.parent

File: tests/java/agents/MCTS.java
package tests.java.agents;

import java.util.ArrayList;
import java.util.Random;
import java.lang.Math;

import tests.java.games.BoardState;
import tests.java.games.Move;

public class MCTS {
    public MCTS() {
    }

    public Move bestMove(Node root, int iterations) {
        for (int i = 0; i < iterations; i++) {
            Node selectedNode = _selection(root);
            Node expandedNode = _expansion(selectedNode);
            int result = _simulation(expandedNode);
            _backpropagation(expandedNode, result);
        }

        Node bestChild = root
            .getChildren()
            .stream()
            .max((child1, child2) ->
                Integer.compare(child1.getVisits(), child2.getVisits())
            )
            .orElseThrow(() -> new RuntimeException("No best child found"));

        return bestChild.getMove();
    }

    private Node _selection(Node root) {
        Node node = root;
        while (node.isFullyExpanded() && !node.isTerminal()) {
            node = node.bestChild(1.41);
        }
        return node;
    }

    private Node _expansion(Node selectedNode) {
        if (selectedNode.isTerminal()) {
            return selectedNode;
        }
        ArrayList<Node> children = selectedNode.getChildren();
        return children
            .stream()
            .filter(child -> child.getVisits() == 0)
            .findAny()
            .orElseThrow(() -> new RuntimeException("No unexplored children"));
    }

    private int _simulation(Node expandedNode) {
        BoardState simulatedGame = expandedNode.getState().copy();
        while (!simulatedGame.isGameOver()) {
            Move move = randomMove(simulatedGame);
            simulatedGame.makeMove(move);
        }
        return simulatedGame.getWinner();
    }

    private void _backpropagation(Node terminalNode, int result) {
        Node node = terminalNode;
        while (node != null) {
            int value = result != node.getState().getPlayer() ? Math.abs(result) : -Math.abs(result);
            node.setValue(node.getValue() + value);
            node.setVisits(node.getVisits() + 1);
            node = node.getParent();
        }
    }

    private Move randomMove(BoardState simulatedGame) {
        ArrayList<Move> legalMoves = simulatedGame.getLegalMoves();
        Random rand = new Random();
        return legalMoves.get(rand.nextInt(legalMoves.size()));
    }
}